cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -118,36 +118,74 @@ private string getValue(Expr e) {
     )
 }
 
+/**
+ * A bitwise `&` expression in which both operands are unsigned, or are effectively
+ * unsigned due to being a non-negative constant.
+ */
+private class UnsignedBitwiseAndExpr extends BitwiseAndExpr {
+  UnsignedBitwiseAndExpr() {
+    (
+      getLeftOperand().getType().getUnderlyingType().(IntegralType).isUnsigned() or
```

Comment:
Does this do the right thing with conversions? For example, does it handle `unsigned z = (unsigned)x & (unsigned)y`, where `x` and `y` are of type `int`?

Reply:
Well spotted - it doesn't, it should be working off the fully converted expression as that's the true input to `&`.  I've pushed some tests and a fix.